third_party/nvfuser/csrc/multigroup_fusion.cpp

```diff
@@ -0,0 +1,150 @@
+
+  // // Add group inputs
+  std::for_each(input_vals.begin(), input_vals.end(), 
+                  [&](auto input){fusion_copy->addInput(original_to_copy_map.clone(input));});
+
+
+  // // Add group outputs
+  std::for_each(output_vals.begin(), output_vals.end(), 
+                  [&](auto output){fusion_copy->addOutput(original_to_copy_map.clone(output));});
+
+  return fusion_copy;
+}
+
+
+MultiGroupFusion::MultiGroupFusion() {
+  Fusion();
+  // tag the fusion to be a multigroup fusion
+  // TODO: can be avoided with using dynamic_cast
+  setActiveMultiGroupFusionBuilder(this);
+}
+
+void MultiGroupFusion::newGroup(bool auto_schedule, ProcessRankType process_rank,
+                                                            c10::Device device)
+{
+  //Stores the new group into the fusion's container
+  groups_.push_back(std::make_shared<Group>(this, auto_schedule,
+                                                    process_rank, device));
+
+  //Set the newly created group as the current group
+  setCurrentGroup(groups_.back());
+}
+
+
+void MultiGroupFusion::newStmt(IrBuilderPasskey, Statement* stmt)
+{
+  if (auto expr = dynamic_cast<Expr*>(stmt)) {
+    auto current_group = getCurrentGroup();
+
+    for (auto input_tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
+      // Check that all inputs required by this new expression
+      //  are defined under the current group's context.
+      TORCH_INTERNAL_ASSERT(
+          current_group->context_tensors.has(input_tv),
+          "tensor input ",
+          input_tv->toString(),
+          " not in context");
+
+      // If we are pulling inputs from other groups, we need
+      //  to mark that as a group input.
+      if (!current_group->internal_tensors.has(input_tv)
+          && !std::count (current_group->input_vals.begin(), current_group->input_vals.end(), input_tv)
+          ){
+        current_group->addInput(input_tv);
+      }
+    }
+
+    for (auto output_tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      // Track defined values under current context
+      current_group->context_tensors.pushBack(output_tv);
+
+      // Track internally defined tensors, i.e. not from context.
+      current_group->internal_tensors.pushBack(output_tv);
+    }
+
+    // stores the current expr in the group's container.
+    current_group->exprs_.push_back(expr);
+  }
+}
+
+void MultiGroupFusion::addGroupOutput(TensorView* tv) {
+  auto group = getCurrentGroup();
+
+  // Check that the given tensor is defined internally
+  //  within the group's context.
+  TORCH_INTERNAL_ASSERT(
+      group->internal_tensors.has(tv), tv->toString(), "not in group");
+
+  // Add the tv to the group outputs.
+  group->addOutput(tv);
+
+  // Add the tv to the global context, since
+  //  it is a group output.
+  context_tensor_map_[tv] = group;
+}
+
+void MultiGroupFusion::addFusionOutput(TensorView* tv) {
```

Comment:
Why does `Fusion::addOutput` need to be wrapped?

Reply:
not strictly needed for now